src/engine/cudaengine.cpp

```diff
@@ -0,0 +1,140 @@
+#include "reffine/engine/cudaengine.h"
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+#include <cassert>
+#include <iostream>
+
+using namespace reffine;
+using namespace std;
+using namespace std::placeholders;
+
+CUDAEngine *CUDAEngine::Init(Module &llmod)
+{
+    static unique_ptr<CUDAEngine> engine;
+
+    if (!engine) {
+        engine = make_unique<CUDAEngine>(llmod);
+
+        InitializeAllTargets();
+        InitializeAllTargetMCs();
+        InitializeAllAsmPrinters();
+        InitializeAllAsmParsers();
+
+        llmod.setTargetTriple("nvptx64-nvidia-cuda");
+        engine->kernel_name = llmod.getName().str();
+    }
+
+    return engine.get();
+}
+
+#define checkCudaErrors(err) __checkCudaErrors(err, __FILE__, __LINE__)
+static void __checkCudaErrors(CUresult err, const char *filename, int line)
+{
+    assert(filename);
+    if (CUDA_SUCCESS != err) {
+        const char *ename = NULL;
+        const CUresult res = cuGetErrorName(err, &ename);
+        fprintf(stderr,
+                "CUDA API Error %04d: \"%s\" from file <%s>, "
+                "line %i.\n",
+                err, ((CUDA_SUCCESS == res) ? ename : "Unknown"), filename,
+                line);
+        exit(err);
+    }
+}
+
+TargetMachine *CUDAEngine::get_target()
+{
+    std::string Error;
+    const llvm::Target *Target =
+        llvm::TargetRegistry::lookupTarget("nvptx64-nvidia-cuda", Error);
+
+    llvm::TargetOptions opt;
+    llvm::TargetMachine *TM =
+        Target->createTargetMachine("nvptx64-nvidia-cuda", "sm_52",
+                                    "+ptx76",  // PTX version
+                                    opt, llvm::Reloc::Static);
+
+    llmod.setDataLayout(TM->createDataLayout());
+
+    return TM;
+}
+
+void CUDAEngine::GeneratePTX()
+{
+    auto TM = get_target();
+
+    llvm::SmallString<1048576> PTXStr;
+    llvm::raw_svector_ostream PTXOS(PTXStr);
+
+    llvm::legacy::PassManager PM;
+    TM->addPassesToEmitFile(PM, PTXOS, nullptr,
+                            llvm::CodeGenFileType::AssemblyFile);
+    PM.run(llmod);
+
+    cout << "Finished generating PTX..." << endl;
+    ptx_str = PTXStr.str().str();
+}
+
+void CUDAEngine::ExecutePTX(void *arg, int len)
```

Comment:
when you implement something, it is important to make sure the they are general enough to handle any arbitrary inputs. In this case, this code can only execute certain type of kernels.

Since the kernel execution is dependent on the query definition, I would suggest to remove this whole function. you can write custom implementations like this in the main.cpp or other example code.